### ISSUE TITLE
Refactor: Extend ipython-sql

### ISSUE THREAD

**Issue by duncandewhurst** [open-contracting/kingfisher-colab#29]
Recently I've been experimenting with [ipython-sql](https://pypi.org/project/ipython-sql/) for connecting notebooks to (non-kingfisher) databases and issuing SQL commands using line and cell [magics](https://ipython.readthedocs.io/en/stable/interactive/magics.html#built-in-magic-commands).

It provides some of the same functionality as kingfisher-colab (`create_connection`, `get_dataframe_from_query` and  `download_dataframe_as_csv`), has some other useful features and makes for a cleaner looking notebook.

We should consider whether to switch to using ipython-sql, which could help to reduce the number of functions we need to maintain in kingfisher-colab.

**Comment by jpmckinney** [comment]
@duncandewhurst If you'd like to convert one of the re-used notebooks to ipython-sql, then we can see what can be cut, and remove those functions from the next version of kingfisher-colab (older notebooks can just install a specific version to avoid having to upgrade).

**Comment by jpmckinney** [comment]
Some methods unique to kingfisher-colab presently require establishing a connection using `create_connection`, but we can see if we can instead just build on ipython-sql.

**Comment by duncandewhurst** [comment]
I haven't had a chance to convert an existing notebook yet, but I've made extensive use of ipython-sql in the [OC4IDS Data Feedback Notebook](https://colab.research.google.com/drive/1V-mhlY1SZpN4pn57TTLn9GUdj6FtAnGr#scrollTo=-Ygw2Mk3Maol&uniqifier=1) and I found it to be very useful.

I came across a [bug](https://github.com/catherinedevlin/ipython-sql/issues/177) with variable substitution in multi-line queries which might cause issues with some of the factored out queries in the OCDS Data Feedback template. However, it should be easy enough to work around by constructing the query string before passing it to the %sql magic and it is still possible to use bind variables to substitute collection ids etc, which is our main use for substitution. 

For the kingfisher-colab functions which include some SQL, if it would be difficult to build on ipython-sql an alternative could be to use Colaboratory's custom code snippets feature (described [here](https://blog.abivin.vn/2019/09/how-to-save-custom-google-colab-snippets.html)).

@mrshll1001 @pindec @odscrachel @yolile @romifz @camilamila - what are your thoughts on moving to using ipython-sql to connect to Kingfisher? Let me know if you'd like a quick demo, or you can check out the [OC4IDS Data Feedback Notebook](https://colab.research.google.com/drive/1V-mhlY1SZpN4pn57TTLn9GUdj6FtAnGr#scrollTo=-Ygw2Mk3Maol&uniqifier=1) to see how it looks

**Comment by romifz** [comment]
I had a look at the notebook and I like the idea of having SQL only cells. I would like to test ipython-sql a little, but I think I would be happy to use it.

**Comment by Camilamila** [comment]
Same, I would be happy to use it.  I had a look at the OC4IDS notebook and found it useful to have SQL only cells.

**Comment by kindly** [comment]
Just a few notes after looking into this.  

I think creating the connection using `create_connection` in this library and also passing on the details to ipython-sql by just creating and `Connection` instance would be easiest. See
https://github.com/catherinedevlin/ipython-sql/blob/master/src/sql/connection.py#L26
This creates a global connection object for the session which the %%sql magic will use.

The other option is extracting the connection details from ipython-sql but this is more complicated to implement and may have side effects not thought about.

This means they both with have own connections but the connection details will be shared, so the user would only have to enter it once. 

Also would be good to create default search paths when working on views for both cases, details of this can be found here:
https://stackoverflow.com/a/49930672

**Comment by pindec** [comment]
+1 for ipython-sql after using it in OC4IDS notebooks.

**Comment by Bjwebb** [comment]
Having looked into this, my preferred approach is to replace all sql calls in kingfisher-colab with calls to ipython-sql. This removes the need for us to pass around connection details.

The ipython-sql code we want to call is in the [SqlMagic class](https://github.com/catherinedevlin/ipython-sql/blob/b24ac6e9410416eafde86ae22fd8d6f34acbe05d/src/sql/magic.py#L34), so we have a choice of setting an instance of that up ourselves or using ipython's magic function call. I think the latter is easier, but it does behave weirdly, for example pulling variables from the local scope, e.g.:

```python
from IPython import get_ipython
ipython = get_ipython() 

def list_collections(source_id):
    """
    Returns, as a data frame, a list of collections with the given source ID.

    :param str source_id: a source ID
    :returns: the results as a data frame
    :rtype: pandas.DataFrame
    """

    sql = """
    SELECT *
    FROM collection
    WHERE source_id = :source_id
    ORDER BY id DESC
    """

    # This inspects locals to find source_id
    return ipython.magic("sql {}".format(sql))
```

Normally I'd avoid this, but I think its maybe okay in a library that only runs in notebooks anyway. It also means we're depending on the advertised api (the magic) rather than how its implemented.

This approach will break backwards compatibility, but my understanding is we're happy to do that (including removing any functions that ipython-sql provides direct alternatives for).

@jpmckinney What do you think?

**Comment by jpmckinney** [comment]
Yes, sounds good! For backwards compatibility, we'll just number the version semantically.

**Comment by jpmckinney** [comment]
Small note: Prefer f-strings for simple substitution: https://ocp-software-handbook.readthedocs.io/en/latest/python/code.html#string-formatting

**Comment by Bjwebb** [comment]
Makes sense, thanks.

Just spotted that with a switch to ipython-sql, we'll also lose [this comment at the top of executed sql](https://github.com/open-contracting/kingfisher-colab/blob/master/ocdskingfishercolab/__init__.py#L142), with a link to the notebook. I'm not sure we can easily recreate this, but probably we can do without it. (Existing use of ipython-sql, such as the OC4IDS notebook also has this problem).

**Comment by jpmckinney** [comment]
That comment has been handy – some analyst queries are extremely inefficient.

In the library, instead of:

```python
return ipython.magic("sql {}".format(sql))
```

Can't we just do:

```
return execute_statement(sql)
```

where an updated `execute_statement` method adds the comment to the SQL passed to `ipython.magic`?

For code in the notebook that doesn't go through kingfisher-colab, I don't know how easy it is to patch ipython-sql.

**Comment by Bjwebb** [comment]
Looks like we should be able to monkeypatch ipython-sql fairly easily, so I'll do that.

**Comment by Bjwebb** [comment]
I've added my work so far to a PR https://github.com/open-contracting/kingfisher-colab/pull/36, but it still needs work on tests.

I've shared a demo of this at https://colab.research.google.com/drive/1cUYY4on72831DPSiQ_JLxJEY2uGTfVrN (not public).

@duncandewhurst Could you have a play around with this, to see if it behaves as expected? You should be able to try this on any notebook, you just need the pip install line from the top.

**Comment by Bjwebb** [comment]
I've released 0.3.0 and update the [demo notebook](https://colab.research.google.com/drive/1cUYY4on72831DPSiQ_JLxJEY2uGTfVrN#scrollTo=C2RbU-iJuyR1).
